File: Tree/leetcode_94.cpp
// Problem Link ====>>https://leetcode.com/problems/binary-tree-inorder-traversal/description/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<int> inorderTraversal(TreeNode *root)
    {
        vector<int> result;
        helper(root, result);
        return result;
    }

    void helper(TreeNode *root, vector<int> &result)
    {
        if (root != nullptr)
        {
            helper(root->left, result);
            result.push_back(root->val);
            helper(root->right, result);
        }
    }
};

int main()
{
    // Create a sample tree for demonstration
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);

    // Create an instance of the Solution class
    Solution solution;

    // Call the inorderTraversal function to get the inorder traversal sequence
    vector<int> result = solution.inorderTraversal(root);

    // Print the inorder traversal sequence
    cout << "Inorder traversal: ";
    for (int num : result)
    {
        cout << num << " ";
    }
    cout << endl;

    // Clean up the tree memory
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}

File: Tree/leetcode_94.py
# Problem Link ====>>https://leetcode.com/problems/binary-tree-inorder-traversal/description/

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        

class Solution(object):
    def inorderTraversal(self, root):
        def helper(root, result):
          if root != None:
            helper(root.left, result)
            result.append(root.val)
            helper(root.right, result)

        result = []
        helper(root, result)
        return result


# Create a sample tree
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

# Create an instance of the Solution class
solution = Solution()

# Call the inorderTraversal function to get the inorder traversal sequence
result = solution.inorderTraversal(root)

# Print the inorder traversal sequence
print("Inorder traversal:", result)
